File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/features/llm/GenerationAnalyticsService.kt
package com.runanywhere.sdk.features.llm

import com.runanywhere.sdk.events.EventPublisher
import com.runanywhere.sdk.foundation.SDKLogger
import com.runanywhere.sdk.foundation.currentTimeMillis
import com.runanywhere.sdk.models.enums.InferenceFramework
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * LLM generation analytics service for tracking generation operations.
 * Mirrors iOS GenerationAnalyticsService exactly.
 *
 * Tracks:
 * - Non-streaming generations (single operation)
 * - Streaming generations (token-by-token with TTFT)
 * - Metrics calculation and aggregation
 *
 * NOTE: Token estimation uses ~4 chars/token approximation (not exact tokenizer count).
 * For precise token counts, use the model's actual tokenizer.
 */
class GenerationAnalyticsService {

    private val logger = SDKLogger("GenerationAnalytics")

    // Active generations
    private val activeGenerations = mutableMapOf<String, GenerationTracker>()

    // Metrics - separated by mode
    private var totalGenerations = 0
    private var streamingGenerations = 0
    private var nonStreamingGenerations = 0
    private var totalTimeToFirstToken = 0.0  // Only for streaming with TTFT recorded
    private var streamingTTFTCount = 0       // Only count TTFT for streaming generations that recorded it
    private var totalTokensPerSecond = 0.0
    private var totalInputTokens = 0
    private var totalOutputTokens = 0
    private val startTime = currentTimeMillis()
    private var lastEventTime: Long? = null

    // MARK: - Types

    private data class GenerationTracker(
        val startTime: Long,
        val modelId: String,
        val isStreaming: Boolean,
        val framework: InferenceFramework,
        var firstTokenTime: Long? = null
    )

    // MARK: - Generation Tracking

    /**
     * Start tracking a non-streaming generation.
     * @param modelId The model being used
     * @param framework The inference framework being used
     * @return A unique generation ID for tracking
     */
    @OptIn(ExperimentalUuidApi::class)
    fun startGeneration(
        modelId: String,
        framework: InferenceFramework = InferenceFramework.LLAMA_CPP
    ): String {
        val id = Uuid.random().toString()

        synchronized(activeGenerations) {
            activeGenerations[id] = GenerationTracker(
                startTime = currentTimeMillis(),
                modelId = modelId,
                isStreaming = false,
                framework = framework
            )
        }

        EventPublisher.track(
            LLMEvent.GenerationStarted(
                generationId = id,
                modelId = modelId,
                prompt = null,  // Privacy: don't log prompts by default
                isStreaming = false,
                framework = framework
            )
        )

        logger.debug("Non-streaming generation started: $id")
        return id
    }

    /**
     * Start tracking a streaming generation.
     * @param modelId The model being used
     * @param framework The inference framework being used
     * @return A unique generation ID for tracking
     */
    @OptIn(ExperimentalUuidApi::class)
    fun startStreamingGeneration(
        modelId: String,
        framework: InferenceFramework = InferenceFramework.LLAMA_CPP
    ): String {
        val id = Uuid.random().toString()

        synchronized(activeGenerations) {
            activeGenerations[id] = GenerationTracker(
                startTime = currentTimeMillis(),
                modelId = modelId,
                isStreaming = true,
                framework = framework
            )
        }

        EventPublisher.track(
            LLMEvent.GenerationStarted(
                generationId = id,
                modelId = modelId,
                prompt = null,  // Privacy: don't log prompts by default
                isStreaming = true,
                framework = framework
            )
        )

        logger.debug("Streaming generation started: $id")
        return id
    }

    /**
     * Track first token (for streaming generations).
     * Only applicable for streaming generations.
     * @param generationId The generation ID from startStreamingGeneration
     */
    fun trackFirstToken(generationId: String) {
        val tracker = synchronized(activeGenerations) {
            activeGenerations[generationId]
        } ?: return

        // Only track for streaming generations
        if (!tracker.isStreaming) {
            logger.warn("trackFirstToken called on non-streaming generation: $generationId")
            return
        }

        // Only track once
        if (tracker.firstTokenTime != null) {
            return
        }

        val now = currentTimeMillis()
        synchronized(activeGenerations) {
            activeGenerations[generationId] = tracker.copy(firstTokenTime = now)
        }

        val latencyMs = (now - tracker.startTime).toDouble()

        EventPublisher.track(
            LLMEvent.FirstToken(
                generationId = generationId,
                latencyMs = latencyMs
            )
        )

        logger.debug("First token: $generationId, latency: ${String.format("%.1f", latencyMs)}ms")
    }

    /**
     * Track streaming update (analytics only, for streaming generations).
     * @param generationId The generation ID from startStreamingGeneration
     * @param tokensGenerated Number of tokens generated so far
     */
    fun trackStreamingUpdate(generationId: String, tokensGenerated: Int) {
        val tracker = synchronized(activeGenerations) {
            activeGenerations[generationId]
        } ?: return

        // Only track for streaming generations
        if (!tracker.isStreaming) {
            return
        }

        EventPublisher.track(
            LLMEvent.StreamingUpdate(
                generationId = generationId,
                tokensGenerated = tokensGenerated
            )
        )
    }

    /**
     * Complete a generation.
     * Works for both streaming and non-streaming generations.
     * @param generationId The generation ID from startGeneration or startStreamingGeneration
     * @param inputTokens Number of input (prompt) tokens
     * @param outputTokens Number of output (completion) tokens
     * @param modelId The model that was used
     */
    fun completeGeneration(
        generationId: String,
        inputTokens: Int,
        outputTokens: Int,
        modelId: String
    ) {
        val tracker = synchronized(activeGenerations) {
            activeGenerations.remove(generationId)
        } ?: return

        val endTime = currentTimeMillis()
        val totalTimeMs = (endTime - tracker.startTime).toDouble()
        val totalTimeSeconds = totalTimeMs / 1000.0

        // Calculate tokens per second
        val tokensPerSecond = if (totalTimeSeconds > 0) {
            outputTokens.toDouble() / totalTimeSeconds
        } else {
            0.0
        }

        // Calculate TTFT for streaming only
        val timeToFirstTokenMs: Double? = if (tracker.isStreaming && tracker.firstTokenTime != null) {
            (tracker.firstTokenTime!! - tracker.startTime).toDouble()
        } else {
            null
        }

        // Update metrics
        synchronized(this) {
            totalGenerations++
            if (tracker.isStreaming) {
                streamingGenerations++
                // Only count TTFT if it was actually recorded
                if (timeToFirstTokenMs != null) {
                    totalTimeToFirstToken += timeToFirstTokenMs
                    streamingTTFTCount++
                }
            } else {
                nonStreamingGenerations++
            }
            totalTokensPerSecond += tokensPerSecond
            totalInputTokens += inputTokens
            totalOutputTokens += outputTokens
            lastEventTime = endTime
        }

        EventPublisher.track(
            LLMEvent.GenerationCompleted(
                generationId = generationId,
                modelId = modelId,
                inputTokens = inputTokens,
                outputTokens = outputTokens,
                durationMs = totalTimeMs,
                tokensPerSecond = tokensPerSecond,
                isStreaming = tracker.isStreaming,
                timeToFirstTokenMs = timeToFirstTokenMs,
                framework = tracker.framework
            )
        )

        logger.debug(
            "Generation completed: $generationId, " +
            "in: $inputTokens, out: $outputTokens, " +
            "time: ${String.format("%.1f", totalTimeMs)}ms, " +
            "speed: ${String.format("%.2f", tokensPerSecond)} tok/s"
        )
    }

    /**
     * Track generation failure.
     * @param generationId The generation ID
     * @param error The error that occurred
     */
    fun trackGenerationFailed(generationId: String, error: Throwable) {
        synchronized(activeGenerations) {
            activeGenerations.remove(generationId)
        }
        lastEventTime = currentTimeMillis()

        EventPublisher.track(
            LLMEvent.GenerationFailed(
                generationId = generationId,
                error = error.message ?: "Unknown error"
            )
        )

        logger.error("Generation failed: $generationId, error: ${error.message}")
    }

    /**
     * Track generation failure with error message.
     * @param generationId The generation ID
     * @param errorMessage The error message
     */
    fun trackGenerationFailed(generationId: String, errorMessage: String) {
        synchronized(activeGenerations) {
            activeGenerations.remove(generationId)
        }
        lastEventTime = currentTimeMillis()

        EventPublisher.track(
            LLMEvent.GenerationFailed(
                generationId = generationId,
                error = errorMessage
            )
        )

        logger.error("Generation failed: $generationId, error: $errorMessage")
    }

    /**
     * Track an error during operations.
     * @param error The error that occurred
     * @param operation Description of the operation that failed
     */
    fun trackError(error: Throwable, operation: String) {
        lastEventTime = currentTimeMillis()
        logger.error("LLM error during $operation: ${error.message}")
        // Error events can be added via a generic ErrorEvent if needed
    }

    // MARK: - Metrics

    /**
     * Get current generation metrics.
     */
    fun getMetrics(): GenerationMetrics {
        return synchronized(this) {
            GenerationMetrics(
                totalEvents = totalGenerations,
                startTime = startTime,
                lastEventTime = lastEventTime,
                totalGenerations = totalGenerations,
                streamingGenerations = streamingGenerations,
                nonStreamingGenerations = nonStreamingGenerations,
                averageTimeToFirstToken = if (streamingTTFTCount > 0) {
                    totalTimeToFirstToken / streamingTTFTCount
                } else {
                    0.0
                },
                averageTokensPerSecond = if (totalGenerations > 0) {
                    totalTokensPerSecond / totalGenerations
                } else {
                    0.0
                },
                totalInputTokens = totalInputTokens,
                totalOutputTokens = totalOutputTokens
            )
        }
    }
}

// MARK: - Generation Metrics

/**
 * Metrics for LLM generation operations.
 * Mirrors iOS GenerationMetrics struct.
 */
data class GenerationMetrics(
    /** Total number of generation events */
    val totalEvents: Int = 0,
    /** When tracking started (epoch millis) */
    val startTime: Long = currentTimeMillis(),
    /** When the last event occurred (epoch millis) */
    val lastEventTime: Long? = null,
    /** Total number of generations completed (streaming + non-streaming) */
    val totalGenerations: Int = 0,
    /** Number of streaming generations */
    val streamingGenerations: Int = 0,
    /** Number of non-streaming generations */
    val nonStreamingGenerations: Int = 0,
    /** Average time to first token in milliseconds (only for streaming generations with TTFT recorded) */
    val averageTimeToFirstToken: Double = 0.0,
    /** Average tokens per second across all generations */
    val averageTokensPerSecond: Double = 0.0,
    /** Total input tokens processed across all generations */
    val totalInputTokens: Int = 0,
    /** Total output tokens generated across all generations */
    val totalOutputTokens: Int = 0
)

File: sdk/runanywhere-flutter/lib/infrastructure/analytics/services/generation_analytics_service.dart
/// LLM Generation analytics service.
/// Tracks generation operations and metrics.
/// Matches iOS GenerationAnalyticsService.
///
/// NOTE: ⚠️ Token estimation uses ~4 chars/token (approximation, not exact tokenizer count).
/// Actual token counts may vary depending on the model's tokenizer and input content.

import 'dart:async';

import '../../../foundation/logging/sdk_logger.dart';
import '../../../core/module_registry.dart' show LLMGenerationResult;

/// LLM analytics service for tracking generation operations.
/// Model lifecycle events (load/unload) are handled by ManagedLifecycle.
///
/// Supports two generation modes:
/// - **Non-streaming** (`generate()`): Synchronous generation, no TTFT tracking
/// - **Streaming** (`generateStream()`): Asynchronous token-by-token generation with TTFT tracking
class GenerationAnalyticsService {
  final SDKLogger _logger = SDKLogger(category: 'GenerationAnalytics');

  /// Active generation operations
  final Map<String, _GenerationTracker> _activeGenerations = {};

  /// Metrics - separated by mode
  int _totalGenerations = 0;
  int _streamingGenerations = 0;
  int _nonStreamingGenerations = 0;
  double _totalTimeToFirstToken = 0;
  int _streamingTTFTCount = 0; // Only count TTFT for streaming generations
  double _totalTokensPerSecond = 0;
  int _totalInputTokens = 0;
  int _totalOutputTokens = 0;
  final DateTime _startTime = DateTime.now();
  DateTime? _lastEventTime;

  GenerationAnalyticsService();

  /// Start tracking a non-streaming generation (generate())
  /// Returns a unique generation ID for tracking.
  String startGeneration({
    required String modelId,
    String? framework,
  }) {
    final id = DateTime.now().microsecondsSinceEpoch.toString();
    _activeGenerations[id] = _GenerationTracker(
      startTime: DateTime.now(),
      isStreaming: false,
      framework: framework,
    );

    _logger.debug('Non-streaming generation started: $id');
    return id;
  }

  /// Start tracking a streaming generation (generateStream())
  /// Returns a unique generation ID for tracking.
  String startStreamingGeneration({
    required String modelId,
    String? framework,
  }) {
    final id = DateTime.now().microsecondsSinceEpoch.toString();
    _activeGenerations[id] = _GenerationTracker(
      startTime: DateTime.now(),
      isStreaming: true,
      framework: framework,
    );

    _logger.debug('Streaming generation started: $id');
    return id;
  }

  /// Track first token for streaming generation (time-to-first-token metric).
  /// Only applicable for streaming generations. Call is ignored for non-streaming.
  void trackFirstToken({required String generationId}) {
    final tracker = _activeGenerations[generationId];
    if (tracker == null || !tracker.isStreaming) {
      // TTFT is only tracked for streaming generations
      return;
    }

    // Only record if not already recorded
    if (tracker.firstTokenTime != null) return;

    final firstTokenTime = DateTime.now();
    tracker.firstTokenTime = firstTokenTime;

    final latencyMs =
        firstTokenTime.difference(tracker.startTime).inMicroseconds / 1000.0;

    _logger.debug(
        'First token received for $generationId: ${latencyMs.toStringAsFixed(1)}ms');
  }

  /// Track streaming update (analytics only).
  /// Only applicable for streaming generations.
  void trackStreamingUpdate({
    required String generationId,
    required int tokensGenerated,
  }) {
    final tracker = _activeGenerations[generationId];
    if (tracker == null || !tracker.isStreaming) {
      return;
    }
    // Additional tracking can be added here
  }

  /// Complete a generation (works for both streaming and non-streaming).
  void completeGeneration({
    required String generationId,
    required int inputTokens,
    required int outputTokens,
    required String modelId,
  }) {
    final tracker = _activeGenerations.remove(generationId);
    if (tracker == null) return;

    final endTime = DateTime.now();
    final totalTimeMs =
        endTime.difference(tracker.startTime).inMicroseconds / 1000.0;
    final totalTimeSec = totalTimeMs / 1000.0;
    final tokensPerSecond =
        totalTimeSec > 0 ? outputTokens / totalTimeSec : 0.0;

    // Calculate TTFT for streaming generations
    double? timeToFirstTokenMs;
    if (tracker.isStreaming && tracker.firstTokenTime != null) {
      final ttftMs =
          tracker.firstTokenTime!.difference(tracker.startTime).inMicroseconds /
              1000.0;
      timeToFirstTokenMs = ttftMs;
      _totalTimeToFirstToken += ttftMs;
      _streamingTTFTCount++;
    }

    // Update metrics
    _totalGenerations++;
    if (tracker.isStreaming) {
      _streamingGenerations++;
    } else {
      _nonStreamingGenerations++;
    }
    _totalTokensPerSecond += tokensPerSecond;
    _totalInputTokens += inputTokens;
    _totalOutputTokens += outputTokens;
    _lastEventTime = endTime;

    final modeStr = tracker.isStreaming ? 'streaming' : 'non-streaming';
    _logger.debug('Generation completed ($modeStr): $generationId');
    if (timeToFirstTokenMs != null) {
      _logger.debug('TTFT: ${timeToFirstTokenMs.toStringAsFixed(1)}ms');
    }
  }

  /// Track generation failure.
  void trackGenerationFailed({
    required String generationId,
    required Object error,
  }) {
    _activeGenerations.remove(generationId);
    _lastEventTime = DateTime.now();
    _logger.warning('Generation failed: $generationId - $error');
  }

  /// Track an error during operations.
  void trackError(Object error, String operation) {
    _lastEventTime = DateTime.now();
    _logger.error('Error in $operation: $error');
  }

  /// Get current generation analytics metrics.
  GenerationMetrics getMetrics() {
    // Average TTFT only counts streaming generations that had TTFT recorded
    final avgTTFT = _streamingTTFTCount > 0
        ? _totalTimeToFirstToken / _streamingTTFTCount
        : 0.0;

    return GenerationMetrics(
      totalEvents: _totalGenerations,
      startTime: _startTime,
      lastEventTime: _lastEventTime,
      totalGenerations: _totalGenerations,
      streamingGenerations: _streamingGenerations,
      nonStreamingGenerations: _nonStreamingGenerations,
      averageTimeToFirstTokenMs: avgTTFT,
      averageTokensPerSecond: _totalGenerations > 0
          ? _totalTokensPerSecond / _totalGenerations
          : 0.0,
      totalInputTokens: _totalInputTokens,
      totalOutputTokens: _totalOutputTokens,
    );
  }
}

/// Internal tracker for generation operations.
class _GenerationTracker {
  final DateTime startTime;
  final bool isStreaming;
  final String? framework;
  DateTime? firstTokenTime;

  _GenerationTracker({
    required this.startTime,
    required this.isStreaming,
    this.framework,
  });
}

/// Generation metrics.
/// Matches iOS GenerationMetrics.
class GenerationMetrics {
  final int totalEvents;
  final DateTime startTime;
  final DateTime? lastEventTime;

  /// Total number of all generations (streaming + non-streaming)
  final int totalGenerations;

  /// Number of streaming generations (generateStream())
  final int streamingGenerations;

  /// Number of non-streaming generations (generate())
  final int nonStreamingGenerations;

  /// Average time to first token in milliseconds (only for streaming generations).
  /// Returns 0 if no streaming generations have completed.
  final double averageTimeToFirstTokenMs;

  /// Average tokens per second across all generations
  final double averageTokensPerSecond;

  /// Total input tokens processed
  final int totalInputTokens;

  /// Total output tokens generated
  final int totalOutputTokens;

  const GenerationMetrics({
    this.totalEvents = 0,
    required this.startTime,
    this.lastEventTime,
    this.totalGenerations = 0,
    this.streamingGenerations = 0,
    this.nonStreamingGenerations = 0,
    this.averageTimeToFirstTokenMs = 0,
    this.averageTokensPerSecond = 0,
    this.totalInputTokens = 0,
    this.totalOutputTokens = 0,
  });
}

/// Streaming metrics collector for TTFT tracking.
/// Used internally by LLMCapability for streaming generations.
/// Matches iOS StreamingMetricsCollector actor pattern.
class StreamingMetricsCollector {
  final String modelId;
  final String generationId;
  final GenerationAnalyticsService analyticsService;
  final String? framework;
  final int promptLength;

  DateTime? _startTime;
  DateTime? _firstTokenTime;
  final StringBuffer _fullText = StringBuffer();
  bool _firstTokenRecorded = false;
  bool _isComplete = false;
  Object? _error;
  Completer<LLMGenerationResult>? _resultCompleter;

  StreamingMetricsCollector({
    required this.modelId,
    required this.generationId,
    required this.analyticsService,
    this.framework,
    required this.promptLength,
  });

  void markStart() {
    _startTime = DateTime.now();
  }

  void recordToken(String token) {
    _fullText.write(token);

    // Track first token for TTFT metric
    if (!_firstTokenRecorded) {
      _firstTokenRecorded = true;
      _firstTokenTime = DateTime.now();
      analyticsService.trackFirstToken(generationId: generationId);
    }
  }

  void markComplete() {
    _isComplete = true;

    // Simple token estimation (~4 chars per token)
    final inputTokens = (promptLength / 4).ceil().clamp(1, promptLength);
    final text = _fullText.toString();
    final outputTokens = (text.length / 4).ceil().clamp(1, text.length);

    analyticsService.completeGeneration(
      generationId: generationId,
      inputTokens: inputTokens,
      outputTokens: outputTokens,
      modelId: modelId,
    );

    _resultCompleter?.complete(_buildResult());
    _resultCompleter = null;
  }

  void markFailed(Object error) {
    _error = error;
    analyticsService.trackGenerationFailed(
      generationId: generationId,
      error: error,
    );

    _resultCompleter?.completeError(error);
    _resultCompleter = null;
  }

  Future<LLMGenerationResult> waitForResult() async {
    if (_isComplete) {
      return _buildResult();
    }

    if (_error != null) {
      throw _error!;
    }

    _resultCompleter ??= Completer<LLMGenerationResult>();
    return _resultCompleter!.future;
  }

  LLMGenerationResult _buildResult() {
    final endTime = DateTime.now();
    final latencyMs = _startTime != null
        ? endTime.difference(_startTime!).inMicroseconds / 1000.0
        : 0.0;

    // Calculate TTFT for streaming
    double? timeToFirstTokenMs;
    if (_startTime != null && _firstTokenTime != null) {
      timeToFirstTokenMs =
          _firstTokenTime!.difference(_startTime!).inMicroseconds / 1000.0;
    }

    // Simple token estimation (~4 chars per token)
    final text = _fullText.toString();
    final inputTokens = (promptLength / 4).ceil().clamp(1, promptLength);
    final outputTokens = (text.length / 4).ceil().clamp(1, text.length);
    final latencySec = latencyMs / 1000.0;
    final tokensPerSecond = latencySec > 0 ? outputTokens / latencySec : 0.0;

    return LLMGenerationResult(
      text: text,
      thinkingContent: null,
      inputTokens: inputTokens,
      tokensUsed: outputTokens,
      modelUsed: modelId,
      latencyMs: latencyMs,
      framework: framework,
      tokensPerSecond: tokensPerSecond,
      timeToFirstTokenMs: timeToFirstTokenMs,
      thinkingTokens: 0,
      responseTokens: outputTokens,
    );
  }
}
